python/lsst/daf/butler/datastores/file_datastore/retrieve_artifacts.py

```diff
@@ -0,0 +1,72 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("determine_destination_for_retrieved_artifact",)
+
+from lsst.resources import ResourcePath, ResourcePathExpression
+
+
+def determine_destination_for_retrieved_artifact(
+    destination_directory: ResourcePath, source_path: ResourcePath, preserve_path: bool
+) -> ResourcePath:
+    """Determine destination path for an artifact retrieved from a datastore.
+
+    Parameters
+    ----------
+    destination_directory : `ResourcePath`
```

Comment:
```suggestion
    destination_directory : `~lsst.resources.ResourcePath`
```
I don't think Sphinx resolves them otherwise because `ResourcePath` is not part of `daf_butler`. It only really matters if this docstring turns up in the sphinx and in the longer term we keep dreaming that sphinx will pick up the type annotations and we can remove the type from here...

Reply:
FYI I asked Jonathan about this at JTM.  The latest version of the documentation tooling does do that, just science pipelines has not upgraded to use it yet.